### PR TITLE
DeForest (2004) implementation by @rubendv

### PR DESCRIPTION
**Pull request by astrofrog**
@rubendv - this includes your implementation of the DeForest (2003) algorithm - and seems to work nicely! Will post some comparison examples here.

cc @Cadair

### CODE REVIEW

File: reproject/high_level.py [hunk 1]
```diff
@@ -43,6 +43,7 @@ def reproject(input_data, output_projection, shape_out=None, projection_type='bi
             * 'biquadratic'
             * 'bicubic'
             * 'flux-conserving'
+            * 'deforest' (C.E. DeForest, Solar Physics 2004)
```
Comment (dpshelio) [inline reproject/high_level.py]: In the PR comment it says DeForest 2003, and here 2004...  Is it this one:
DOI: [10.1023/B:SOLA.0000021743.24248.b0](http://dx.doi.org/10.1023/B:SOLA.0000021743.24248.b0)
Maybe adding the DOI number to the docstring would help for people to find the paper?

@astrofrog great job by the way!

**Comment by prateekiiest** [comment]
what needs to be updated in this PR as part of the sunkit project?
can anyone enlighten me upon that

**Comment by Cadair** [comment]
@onoddil or this.

**Comment by astrofrog** [comment]
I've revived this PR after 5 years!

@rubendv - I found a couple of bugs in the Cython implementation that you might want to include in your version - see this commit in particular: https://github.com/astropy/reproject/pull/52/commits/3941b474e82d4305bdfd3bf40d812978ab0ad3c4

This should in principle be fully functional. A few caveats though:

* For now this only works in 2D, but there's actually no reason this couldn't be extended to more dimensions
* I haven't added the round-tripping checks which are important for solar data, but should be easy to add
* This needs tests, docs, and a changelog entry :)

@Cadair - wanna take this for a spin?

File: docs/celestial.rst [hunk 2]
```diff
@@ -99,39 +120,146 @@ We can create a new header using a Gnomonic projection::
...
+
+    >>> from reproject import reproject_adaptive
+
+In addition to the arguments described in :ref:`common`, the order of the
+interpolation to use when sampling values in the input map can be controlled by
+setting the ``order=`` argument to either an integer or a string giving the
+order of the interpolation. Supported strings include:
+
+* ``'nearest-neighbor'``: zeroth order interpolation
+* ``'bilinear'``: fisst order interpolation
```
Comment (Cadair) [inline docs/celestial.rst]:
```suggestion
* ``'bilinear'``: first order interpolation
```

File: reproject/adaptive/high_level.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    -------
+    array_new : `~numpy.ndarray`
+        The reprojected array
+    footprint : `~numpy.ndarray`
+        Footprint of the input array in the output array. Values of 0 indicate
+        no coverage or valid values in the input image, while values of 1
+        indicate valid values.
+    """
+
+    # TODO: add support for output_array and return_footprint
```
Comment (Cadair) [inline reproject/adaptive/high_level.py]:
```suggestion
    # TODO: add support for output_array
```

**Comment by rubendv** [comment]
Great to see that this is useful for you guys! I'm no longer in research myself, so I'm not really maintaining my version any more due to lack of time, but thanks for the heads up @astrofrog!

File: reproject/mosaicking/tests/test_coadd.py [hunk 4]
```diff
@@ -164,3 +171,80 @@ def test_coadd_background_matching_with_nan(self, reproject_function):
...
+    from sunpy.map import Map, all_coordinates_from_map
+
+    # Load in three images from different viewpoints around the Sun
+    filenames = ['secchi_l0_a.fits', 'aia_171_level1.fits', 'secchi_l0_b.fits']
+    maps = [Map(os.path.join(DATA, f)) for f in filenames]
+
+    # Produce weight maps that are centered on the solar disk and go to zero at the edges
+    coordinates = tuple(map(all_coordinates_from_map, maps))
+    input_weights = [coord.transform_to("heliocentric").z.value for coord in coordinates]
+    input_weights = [(w / np.nanmax(w)) ** 4 for w in input_weights]
```
Comment (Cadair) [inline reproject/mosaicking/tests/test_coadd.py]: four! :stuck_out_tongue:

Comment (astrofrog) [reply reproject/mosaicking/tests/test_coadd.py]: oops didn't actually mean that, it was during testing. But no harm I guess since it's just a test!

**Comment by Cadair** [comment]
YEAH!! Thanks @astrofrog

**Comment by bsipocz** [comment]
@astrofrog - CI is double running on PRs if they are opened on branches from the main repo. It can be worked around if needed, but better to use branches on forks.

**Comment by astrofrog** [comment]
@bsipocz - I agree, but this PR was started back in 2014 when the repo was under my user hence the use of a branch on the main repo here ;)

**Comment by bsipocz** [comment]
Oh, I haven't notice that one. Good job to finish up such an old one 😄

**Pull request merged**